### PR TITLE
Misc paint setup

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/interface/viewport.c [hunk 1]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
 {
-	RCT2_CALLPROC_X(0x00672AC9, misc->unknown.x, imageDirection, misc->unknown.y, misc->unknown.z, (int)misc, 0, 0);
+	rct_drawpixelinfo * dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+
+	switch (misc->steam_particle.misc_identifier) {
+		case SPRITE_MISC_STEAM_PARTICLE: // 0
+		{
+			uint32 imageId = 22637 + (misc->steam_particle.var_26 / 256);
+			sub_98196C(imageId, 0, 0, 1, 1, 0, misc->unknown.z, get_current_rotation());
+		}
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: your missing your `break;` after each of these switch cases.

File: src/interface/viewport.c [hunk 2]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+
+			rct_money_effect moneyEffect = misc->money_effect;
+			money32 amount = moneyEffect.value;
+			uint16 stringId = 1388; // {OUTLINE}{GREEN}+ {CURRENCY}
+			if (amount < 0) {
+				amount = -amount;
+				stringId = 1389; // {OUTLINE}{RED}- {CURRENCY}
+			}
+
+			sub_685EBC(amount, stringId, moneyEffect.y, moneyEffect.z, money_wave[moneyEffect.wiggle], moneyEffect.offset_x, get_current_rotation());
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: shouldn't this be `&money_wave[moneyEffect.wiggle]` ?

File: src/interface/viewport.c [hunk 3]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+				return;
+			}
+
+			rct_crashed_vehicle_particle particle = misc->crashed_vehicle_particle;
+			uint32 imageId = vehicle_particle_base_sprites[particle.var_2E] + particle.var_26 / 256;
+			imageId = imageId | (particle.colour[0] << 19) | (particle.colour[1] << 24) | 0xA0000000;
+			sub_98196C(imageId, 0, 0, 1, 1, 0, misc->unknown.z, get_current_rotation());
+		}
+
+		case SPRITE_MISC_3: // 3
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Can you rename MISC_3 as MISC_EXPLOSION_CLOUD

File: src/interface/viewport.c [hunk 4]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+		{
+			// Explosion cloud
+			uint32 imageId = 22878 + (misc->unknown.var_26 / 256);
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: var_26 is probably a frame number of the animation. Could you label it.

File: src/interface/viewport.c [hunk 5]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+		case SPRITE_MISC_JUMPING_FOUNTAIN_WATER: // 6
+		case SPRITE_MISC_JUMPING_FOUNTAIN_SNOW: // 9
+		{
+			if (dpi->zoom_level != 0) {
+				return;
+			}
+
+			rct_jumping_fountain jumpingFountain = misc->jumping_fountain;
+
+			uint16 height = jumpingFountain.y + 6;
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: `z` surely not `y`.

File: src/interface/viewport.c [hunk 6]
```diff
@@ -1259,13 +1259,191 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+		{
+			rct_balloon balloon = misc->balloon;
+
+			uint32 imageId = 22651 + ((balloon.var_26 / 256) & 7);
+			if (balloon.popped != 0) {
+				imageId += 8;
+			}
+
+			imageId = imageId | (balloon.colour << 19) | 0x20000000;
+			sub_98196C(imageId, 0, 0, 1, 1, 0, balloon.z, get_current_rotation());
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: `balloon.z + 6`

**Comment by marijnvdwerf** [comment]
`MISC_3`/`MISC_EXPLOSION_CLOUD`:

**Comment by marijnvdwerf** [comment]
`MISC_5`
image

**Comment by Gymnasiast** [comment]
Now, I have played RCT since 2000 and I haven't got the foggiest idea as to where `MISC_5` is used.

**Comment by marijnvdwerf** [comment]
@Gymnasiast They seem to be used together: 
scr23

**Comment by Gymnasiast** [comment]
I just saw that in the code. And somehow I never noticed that 'flare' in 16 years. :?

**Comment by duncanspumpkin** [comment]
Can you call it explosion_2 or flare or something, anything is better than MISC_5

File: src/interface/viewport.c [hunk 7]
```diff
@@ -1259,13 +1259,199 @@ void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
...
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
+	0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1
```
Comment (IntelOrca) [inline src/interface/viewport.c]: It looks like you could just divide the index by 22 and cut the array size down dramatically.

Comment (marijnvdwerf) [inline src/interface/viewport.c]: @IntelOrca I think you mean the modulus operator?

Comment (IntelOrca) [inline src/interface/viewport.c]: Modulo operator.

**Comment by marijnvdwerf** [comment]
Build is broken intentionally. Depends on #3438.

**Pull request merged**